[quant][graphmode][fx] Add reference quantized linear module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64135
* #63828
* __->__ #63627

Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

jerryzh168 committed:
[quant][graphmode][fx] Add reference quantized linear module

Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30504750).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

vkuzo commented on torch/nn/intrinsic/quantized/_reference/modules/linear_relu.py:
```diff
@@ -11,17 +11,13 @@ def __init__(
             bias=True,
-            dtype=torch.qint8):
-        super().__init__(in_features, out_features, bias, dtype)
+            weight_qparams=None):
```
any chance we could add a typehint?

vkuzo commented on torch/nn/quantized/_reference/modules/utils.py:
```diff
@@ -0,0 +1,61 @@
+    assert module.weight_qscheme in [None, torch.per_tensor_affine, torch.per_channel_affine], \
+    Exception(f"qscheme: {module.weight_qscheme} is not support in reference quantized linear module")
+    if module.weight_qscheme is not None:
+        module.register_buffer("weight_scale", torch.tensor(weight_qparams["scale"]))
+        module.register_buffer("weight_zero_point", torch.tensor(weight_qparams["zero_point"]))
```
could we move the buffer registration to the actual modules?  I understand that it might decrease code reuse, but it's not intuitive if buffers are created outside of the module code in a util function.  I think in this case it's better to optimize for readability and the code being obvious.

A motivating example where the current code might be confusing is if a user is looking to see which parameters and buffers are defined on the class.  I'd expect most users to assume that they can find all of the parameters and buffers in the class definitions on the inheritance chain.

jerryzh168 replied:
yeah sure

vkuzo commented on torch/nn/quantized/_reference/modules/utils.py:
```diff
@@ -0,0 +1,61 @@
+            module.register_buffer("weight_axis", torch.tensor(0))
+
+
+def _quantize_and_dequantize_weight(weight, weight_qscheme, weight_dtype, weight_scale, weight_zero_point, weight_axis):
```
could we add typehints?

vkuzo reviewed (commented):
looks great, had one comment about not registering buffers in util functions

vkuzo commented on torch/nn/quantized/_reference/modules/utils.py:
```diff
@@ -0,0 +1,61 @@
+        if weight_qscheme == torch.per_channel_affine:
+            destination[prefix + "weight_axis"] = weight_axis
+
+def _load_weight_qparams(module, state_dict):
```
might be good to refactor this function to be stateless and return the results, and attach the results to the module in the caller.  This way the reader doesn't have to navigate to this function to see which side effects it has.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference quantized linear module"


Summary:
Added reference quantized linear module for the custom backend flow, the reference quantized module will
have the following code:
```
        w(float) -- quant - dequant \
        x(float) ------------- F.linear ---
```
In the full model, we will see
```
        w(float) -- quant - *dequant \
        x -- quant --- *dequant --  *F.linear --- *quant - dequant
```
and the backend should be able to fuse the ops with `*` into a quantized linear

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_conv_linear_reference

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30504750](https://our.internmc.facebook.com/intern/diff/D30504750)

[ghstack-poisoned]
